### PR TITLE
NAS-129734 / 24.10 / Prepare directory services cache for IPA changes

### PR DESCRIPTION
**Pull request by anodos325**
This commit creates a generic directory-service agnostic user and group cache in perparation for adding IPA directory service integration. It also includes opportunistic API cleanups by removing the TDB plugin that was initially written for clustering support in favor of shifting TDB wrappers into a utils file with direct CI tests.

### CODE REVIEW

File: src/middlewared/middlewared/plugins/directoryservices_/util_cache.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+        return nss_entries
+
+    def _get_entries_for_cache(
+        self,
+        idmap_ctx: idmap_winbind.WBClient | idmap_sss.SSSClient | None,
+        nss_module: NssModule,
+        entry_type: IDType,
+        dom_by_sid: dict
+    ) -> list:
```
Comment (mgrimesix) [inline src/middlewared/middlewared/plugins/directoryservices_/util_cache.py:167]: Adding type hints and return types gives more opportunity for Flake8 to find things to complain about.

> Return type of generator function must be compatible with "Generator[Any, Any, Any]"
>   "Generator[Any, Any, Any]" is incompatible with "list"

File: src/middlewared/middlewared/plugins/account.py [hunk 2]
```diff
@@ -292,7 +292,7 @@ async def query(self, filters, options):
         if dssearch:
```
Comment (mgrimesix) [inline src/middlewared/middlewared/plugins/account.py:292]: While not part of your changes, we could clean this module up by changing the import on line 18 from:
`from middlewared.api.current import *`
to:
```
from middlewared.api.current import (
    UserEntry, UserCreateArgs, UserCreateResult, UserUpdateArgs, UserUpdateResult
)
```

File: src/middlewared/middlewared/plugins/ldap.py [hunk 3]
```diff
@@ -945,75 +945,7 @@ async def __stop(self, job):
         await self.middleware.call('etc.generate', 'pam')
```
Comment (mgrimesix) [inline src/middlewared/middlewared/plugins/ldap.py:945]: Might as well clean this up too.
**Imports:**
    Line 18: Remove this line `middlewared.utils.nss.pwd` and `grp` are unused.
    Line 19: Remove this line `NssModule` is unused.

**General Flake8 complaints:**
    Line 676: Missing whitespace after keyword (col 16)

**Typo that should be fixed:**  <----------
    Line 737: Missing 'self'
        (This is the `kinit` function, not sure how it's working!)
        In that same function `krb5ccache` and `ldap_config` are undefined.  Possibly related to the missing `self`.

File: src/middlewared/middlewared/pytest/unit/utils/test_tdb.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
+import gc
+import os
+import pytest
+
+from base64 import b64encode, b64decode
```
Comment (mgrimesix) [inline src/middlewared/middlewared/pytest/unit/utils/test_tdb.py:5]: b64decode is not used

File: src/middlewared/middlewared/pytest/unit/utils/test_tdb.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+from contextlib import closing
+from middlewared.plugins.sysdataset import SYSDATASET_PATH
+from middlewared.utils.tdb import (
+    close_sysdataset_tdb_handles,
+    get_tdb_handle,
+    TDBBatchAction,
+    TDBBatchOperation,
+    TDBDataType,
+    TDBHandle,
+    TDBError,
```
Comment (mgrimesix) [inline src/middlewared/middlewared/pytest/unit/utils/test_tdb.py:15]: TDBError is not used.

**Review by mgrimesix — CHANGES_REQUESTED** [review]
There appears to be a typo, not part of this PR, but should be fixed.

File: tests/api2/test_040_ad_user_group_cache.py [hunk 6]
```diff
@@ -1,165 +1,127 @@
...
 
-    results = GET('/user', payload={
-        'query-filters': [['local', '=', False]],
-        'query-options': {'extra': {"search_dscache": True}},
-    })
-    assert results.status_code == 200, results.text
-    assert len(results.json()) > 0, results.text
-    cache_names = [x['username'] for x in results.json()]
+    cache_names = set([x['username'] for x in call(
+       'user.query',
```
Comment (mgrimesix) [inline tests/api2/test_040_ad_user_group_cache.py:63]: More vision problems.   Indentation is only 3.
This same indentation issue is repeated at lines: 82, 108, 116, 142, 150, 186 and 194.

File: tests/api2/test_040_ad_user_group_cache.py [hunk 7]
```diff
@@ -1,165 +1,127 @@
...
-
-@pytest.mark.dependency(name="LAZY_INITIALIZATION_BY_NAME")
-def test_10_check_lazy_initialization_of_users_and_groups_by_name(request):
+    users = set([x['username'] for x in call(
+       'user.query',
+       [['local', '=', False]],
+       {'extra': {'search_dscache': True}}
+    )])
+
+    assert len(users) > 0
```
Comment (mgrimesix) [inline tests/api2/test_040_ad_user_group_cache.py:113]: Maybe also confirm actual unique names, not just that some names exist or do we have an expectation on the number of entries that we can test against?

**Review by mgrimesix — CHANGES_REQUESTED** [review]

File: src/middlewared/middlewared/utils/tdb.py [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+        """
+        Retrieve the specified key
+
+        Returns:
+            dict if TDBDatatype is JSON
+            str if TDBDatatype is BYTES or STRING
+
+        Raises:
+            MatchNotFound
+            RuntimeError
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/utils/tdb.py:140]: I've added ValueError as possible option. tdb raises RuntimeError internally.

**Review by mgrimesix — APPROVED** [review]

File: src/middlewared/middlewared/pytest/unit/utils/test_tdb.py [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+    # now fetch one and delete the other two
+    batched_ops = [
+        TDBBatchOperation(TDBBatchAction.GET, key='test_key1', value=data1),
+        TDBBatchOperation(TDBBatchAction.DEL, key='test_key2'),
+        TDBBatchOperation(TDBBatchAction.DEL, key='test_key3'),
+    ]
+    out = hdl.batch_op(batched_ops)
+    assert out['test_key1'] == data1
+
```
Comment (bmeagherix) [inline src/middlewared/middlewared/pytest/unit/utils/test_tdb.py:107]: Maybe add a 
```
    assert len(list(hdl.entries())) == 1
```

File: tests/api2/test_040_ad_user_group_cache.py [hunk 10]
```diff
@@ -1,165 +1,127 @@
...
-    results = GET('/directoryservices/cache_refresh/')
-    assert results.status_code == 200, results.text
-    if results.status_code == 200:
-        refresh_job = results.json()
-        job_status = wait_on_job(refresh_job, 180)
-        assert job_status['state'] == 'SUCCESS', str(job_status['results'])
-        if job_status['state'] == 'SUCCESS':
-            rebuild_ok = True
+    # directoryservices.cache_refresh job causes us to rebuild / refresh LDAP / AD users.
+    call('directoryservices.cache.refresh', job=True)
```
Comment (bmeagherix) [inline tests/api2/test_040_ad_user_group_cache.py:105]: Do we need to wait for this job to complete before doing the subsequent queries?

File: src/middlewared/middlewared/plugins/activedirectory.py [hunk 11]
```diff
@@ -890,7 +891,6 @@ async def __stop(self, job, config):
         await self.middleware.call('etc.generate', 'nss')
         await self.set_state(DSStatus['DISABLED'].name)
         job.set_progress(60, 'clearing caches.')
-        await self.middleware.call('service.stop', 'dscache')
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/plugins/activedirectory.py:893]: Interesting idea. I'll need to double-check behavior of iterpwd / itergrp when we reconfigure nsswitch under it (I think it'll fail naturally causing the cache fill to terminate with error).

[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/plugins/activedirectory.py:893]: Nope. Looks like it keeps filling. Good idea on killing the fill job 👍

File: tests/api2/test_040_ad_user_group_cache.py (hunk 10 above)
Comment (anodos325) [reply tests/api2/test_040_ad_user_group_cache.py:105]: `job=True` automatically waits.

**Pull request merged**